File: HKDConvertorJava_MVI/app/src/main/java/com/example/hkdconvertorjava/viewstate/MainViewState.java
package com.example.hkdconvertorjava.viewstate;

// Represents different states that the View can be in
public class MainViewState {

    // Idle state when no action is being performed
    public static class Idle extends MainViewState {}

    // Loading state when a conversion is in progress
    public static class Loading extends MainViewState {}

    // Success state when the conversion is successful
    public static class Success extends MainViewState {
        public final String result;  // Holds the conversion result

        // Constructor to initialize the result
        public Success(String result) {
            this.result = result;
        }
    }

    // Error state when an error occurs during the conversion process
    public static class Error extends MainViewState {
        public final String error;  // Holds the error message

        // Constructor to initialize the error message
        public Error(String error) {
            this.error = error;
        }
    }
}

File: HKDConvertorKotlin_MVI/app/src/main/java/com/example/hkdconvertorkotlin/MainViewState.kt
package com.example.hkdconvertorkotlin

// This sealed class represents all possible states of the UI.
// It ensures that the UI can only be in one of the defined states.
sealed class MainViewState {
    // Idle state: when nothing is happening (initial state).
    object Idle : MainViewState()

    // Loading state: shown when the currency conversion is in progress.
    object Loading : MainViewState()

    // Success state: when the conversion finishes successfully.
    // The result will be displayed in the UI.
    data class Success(val result: String) : MainViewState()

    // Error state: when an error occurs during the conversion process.
    // It carries the error message to be shown in the UI.
    data class Error(val error: String) : MainViewState()
}